File: app/src/main/java/com/denchic45/kts/uivalidator/Validation.java
package com.denchic45.kts.uivalidator;

import androidx.lifecycle.MutableLiveData;

import com.denchic45.kts.SingleLiveData;
import com.denchic45.kts.uivalidator.validatorlistener.ActionValidationListener;
import com.denchic45.kts.uivalidator.validatorlistener.FieldMessageIdLiveDataValidationListener;
import com.denchic45.kts.uivalidator.validatorlistener.FieldMessageLiveDataValidationListener;
import com.denchic45.kts.uivalidator.validatorlistener.MessageIdLiveDataValidationListener;
import com.denchic45.kts.uivalidator.validatorlistener.MessageLiveDataValidationListener;
import com.denchic45.kts.uivalidator.validatorlistener.ValidationListener;

import org.jetbrains.annotations.NotNull;

import java.util.function.Consumer;

import kotlin.Pair;

public class Validation {

    private final Rule[] rules;
    private Runnable errorRunnable;
    private Consumer<String> messageConsumer;
    private Consumer<Integer> resIdConsumer;
    private Runnable successRunnable;
    private Runnable runResetLiveData;

    private ValidationListener validationListener = ValidationListener.EMPTY;

    public Validation(Rule... rules) {
        this.rules = rules;
    }


    public boolean validate() {
        return isValid();
    }

    public Validation onErrorRun(Runnable runnable) {
        errorRunnable = runnable;
        return this;
    }

    public Validation onErrorSendMessage(Consumer<String> consumer) {
        messageConsumer = consumer;
        return this;
    }

    public Validation onErrorSendMessage(int fieldResId, Consumer<String> consumer) {
        messageConsumer = consumer;
        return this;
    }

    public Validation onErrorSendIdRes(Consumer<Integer> consumer) {
        resIdConsumer = consumer;
        return this;
    }

    public Validation onSuccessRun(Runnable runnable) {
        successRunnable = runnable;
        return this;
    }

    public Validation onSuccessResetMessage(@NotNull SingleLiveData<String> liveData) {
        runResetLiveData = () -> liveData.setValue(null);
        return this;
    }

    public Validation onSuccessResetResId(@NotNull SingleLiveData<Integer> liveData) {
        runResetLiveData = () -> liveData.setValue(null);
        return this;
    }

    public Validation onSuccessResetMessage(@NotNull SingleLiveData<Pair<Integer, String>> liveData, int fieldResId) {
        runResetLiveData = () -> liveData.setValue(new Pair<>(fieldResId, null));
        return this;
    }

    public Validation onSuccessResetResId(@NotNull SingleLiveData<Pair<Integer, Integer>> liveData, int fieldResId) {
        runResetLiveData = () -> liveData.setValue(new Pair<>(fieldResId, null));
        return this;
    }

    public Validation sendMessageResult(ValidationListener validationListener) {
        this.validationListener = validationListener;
        return this;
    }

    public Validation sendMessageResult(MutableLiveData<String> mutableLiveData) {
        validationListener = new MessageLiveDataValidationListener(mutableLiveData);
        return this;
    }

    public Validation sendMessageResult(int id, MutableLiveData<Pair<Integer, String>> mutableLiveData) {
        validationListener = new FieldMessageLiveDataValidationListener(id, mutableLiveData);
        return this;
    }

    public Validation sendMessageIdResult(int id, MutableLiveData<Pair<Integer, Integer>> mutableLiveData) {
        validationListener = new FieldMessageIdLiveDataValidationListener(id, mutableLiveData);
        return this;
    }

    public Validation sendMessageIdResult(MutableLiveData<Integer> mutableLiveData) {
        validationListener = new MessageIdLiveDataValidationListener(mutableLiveData);
        return this;
    }

    public Validation sendActionResult(Consumer<Rule> errorAction, Runnable successAction) {
        validationListener = new ActionValidationListener(errorAction, successAction);
        return this;
    }

    private boolean isValid() {
        for (Rule rule : rules) {
            validationListener.run(rule);
            if (!rule.isValid()) {
                onPostError(rule);
                return false;
            }
        }
        onPostSuccess();
        return true;
    }

    protected void onPostSuccess() {
        if (successRunnable != null)
            successRunnable.run();
        if (runResetLiveData != null)
            runResetLiveData.run();
    }

    protected void onPostError(Rule rule) {
        if (errorRunnable != null)
            errorRunnable.run();
        if (messageConsumer != null)
            messageConsumer.accept(rule.getErrorMessage());
        if (resIdConsumer != null)
            resIdConsumer.accept(rule.getErrorResId());
    }


}

File: app/src/main/java/com/denchic45/kts/uivalidator/Validation.kt
package com.denchic45.kts.uivalidator

import androidx.lifecycle.MutableLiveData
import com.denchic45.kts.SingleLiveData
import com.denchic45.kts.uivalidator.validatorlistener.*
import kotlinx.coroutines.flow.MutableSharedFlow
import java.util.function.Consumer

class Validation(private vararg val rules: Rule) {
    private var errorRunnable: Runnable? = null
    private var messageConsumer: Consumer<String>? = null
    private var resIdConsumer: Consumer<Int>? = null
    private var successRunnable: Runnable? = null
    private var runResetLiveData: Runnable? = null
    private var validationListener = ValidationListener.EMPTY
    fun validate(): Boolean {
        return isValid
    }

    fun onErrorRun(runnable: Runnable): Validation {
        errorRunnable = runnable
        return this
    }

    fun onErrorSendMessage(consumer: Consumer<String>): Validation {
        messageConsumer = consumer
        return this
    }

    fun onErrorSendMessage(fieldResId: Int, consumer: Consumer<String>): Validation {
        messageConsumer = consumer
        return this
    }

    fun onErrorSendIdRes(consumer: Consumer<Int>): Validation {
        resIdConsumer = consumer
        return this
    }

    fun onSuccessRun(runnable: Runnable): Validation {
        successRunnable = runnable
        return this
    }

    fun onSuccessResetMessage(liveData: SingleLiveData<String?>): Validation {
        runResetLiveData = Runnable { liveData.setValue(null) }
        return this
    }

    fun onSuccessResetResId(liveData: SingleLiveData<Int?>): Validation {
        runResetLiveData = Runnable { liveData.setValue(null) }
        return this
    }

    fun onSuccessResetMessage(
        liveData: SingleLiveData<Pair<Int, String?>>,
        fieldResId: Int
    ): Validation {
        runResetLiveData = Runnable { liveData.setValue(Pair<Int, String?>(fieldResId, null)) }
        return this
    }

    fun onSuccessResetResId(
        liveData: SingleLiveData<Pair<Int, Int?>>,
        fieldResId: Int
    ): Validation {
        runResetLiveData = Runnable { liveData.setValue(Pair<Int, Int?>(fieldResId, null)) }
        return this
    }

    fun sendMessageResult(validationListener: ValidationListener): Validation {
        this.validationListener = validationListener
        return this
    }

    fun sendMessageResult(mutableLiveData: MutableLiveData<String?>): Validation {
        validationListener = MessageLiveDataValidationListener(mutableLiveData)
        return this
    }

    fun sendMessageIdResult(
        id: Int,
        mutableLiveData: MutableLiveData<Pair<Int, Int?>>
    ): Validation {
        validationListener = FieldMessageIdLiveDataValidationListener(id, mutableLiveData)
        return this
    }

    fun sendMessageResult(
        id: Int,
        mutableLiveData: MutableLiveData<Pair<Int, String?>>
    ): Validation {
        validationListener = FieldMessageLiveDataValidationListener(id, mutableLiveData)
        return this
    }

    fun sendMessageIdResult(mutableLiveData: MutableLiveData<Int>): Validation {
        validationListener = MessageIdLiveDataValidationListener(mutableLiveData)
        return this
    }

    fun sendMessageIdResult(mutableSharedFlow: MutableSharedFlow<Int>): Validation {
        validationListener = MessageIdSharedFlowValidationListener(mutableSharedFlow)
        return this
    }

    fun sendActionResult(errorAction: Consumer<Rule?>?, successAction: Runnable?): Validation {
        validationListener = ActionValidationListener(errorAction, successAction)
        return this
    }

    fun sendMessageResult(sharedFlow: MutableSharedFlow<String>): Validation {
        validationListener = MessageSharedFlowValidationListener(sharedFlow)
        return this
    }

    private val isValid: Boolean
        private get() {
            for (rule in rules) {
                validationListener.run(rule)
                if (!rule.isValid) {
                    onPostError(rule)
                    return false
                }
            }
            onPostSuccess()
            return true
        }

    protected fun onPostSuccess() {
        if (successRunnable != null) successRunnable!!.run()
        if (runResetLiveData != null) runResetLiveData!!.run()
    }

    protected fun onPostError(rule: Rule) {
        if (errorRunnable != null) errorRunnable!!.run()
        if (messageConsumer != null) messageConsumer!!.accept(rule.errorMessage)
        if (resIdConsumer != null) resIdConsumer!!.accept(rule.errorResId)
    }

}